specifyweb/workbench/upload/treerecord.py

```diff
@@ -19,33 +19,138 @@
 
 logger = logging.getLogger(__name__)
 
+class TreeRank:
```

Comment:
Can you combine this or get rid of this, in any way, possibly using `scopingAttrs`? Try to keep props usage to minimal, and only if none of others workout.

Reply:
At least make it namedtuple. It's not about "obsession with readonly". If I want to refactor this, I'd need to assume that you are sneakily mutating something. make stuff easier for others, make it named tuple.